core/templates/pages/library-page/search-bar/search-bar.component.spec.ts

```diff
@@ -0,0 +1,239 @@
+      $location: $location,
+      $rootScope: $rootScope,
+      $scope: $scope,
+      ConstructTranslationIdsService: constructTranslationIdsService,
+      I18nLanguageCodeService: i18nLanguageCodeService
+    });
+
+    // This approach was choosen because spyOn() doesn't work on properties
+    // that doesn't have a get access type.
+    // Without this approach the test will fail because it'll throw
+    // 'Property classroomBackendApiService does not have access type get'
+    // or 'Property classroomBackendApiService does not have access type set'
+    // error.
+    Object.defineProperty(ctrl, 'classroomBackendApiService', {
+      get: () => undefined,
+      set: () => {}
+    });
+    spyOnProperty(ctrl, 'classroomBackendApiService').and.returnValue(
+      classroomBackendApiService);
+    spyOnProperty(
+      classroomBackendApiService,
+      'onInitializeTranslation').and.returnValue(initTranslationEmitter);
+    spyOnProperty(
+      i18nLanguageCodeService,
+      'onPreferredLanguageCodesLoaded').and.returnValue(
+      preferredLanguageCodesLoadedEmitter);
+    ctrl.$onInit();
+  }));
+
+  afterEach(function() {
+    ctrl.$onDestroy();
+  });
+
+  it('should initialize controller properties after its initialization',
+    function() {
+      expect(Object.keys(ctrl.selectionDetails)).toContain('categories');
+      expect(Object.keys(ctrl.selectionDetails)).toContain('languageCodes');
+
+      expect(ctrl.translationData).toEqual({
+        categoriesCount: 0,
+        languagesCount: 0
+      });
+    });
+
+  it('should get placeholder and button text translation when translation' +
+    ' is initialized', function() {
+    initTranslationEmitter.emit();
+
+    expect(ctrl.searchBarPlaceholder).toBe('I18N_LIBRARY_SEARCH_PLACEHOLDER');
+    expect(ctrl.categoryButtonText).toBe('I18N_LIBRARY_ALL_CATEGORIES');
+    expect(ctrl.languageButtonText).toBe('I18N_LIBRARY_ALL_LANGUAGES');
+  });
+
+  it('should search for all content in given languages', function() {
+    expect(ctrl.isSearchInProgress()).toBe(false);
+
+    mockWindow.location.pathname = '/search/find';
+    preferredLanguageCodesLoadedEmitter.emit(['en', 'es', 'hi']);
+    expect(ctrl.isSearchInProgress()).toBe(true);
+
+    expect(ctrl.translationData).toEqual({
+      categoriesCount: 0,
+      languagesCount: 3
+    });
+    expect($location.url()).toBe(
+      '/find?q=&language_code=(%22en%22%20OR%20%22es%22%20OR%20%22hi%22)');
+
+    preferredLanguageCodesLoadedEmitter.emit(['en']);
+
+    expect(ctrl.translationData).toEqual({
+      categoriesCount: 0,
+      languagesCount: 2
+    });
+    expect($location.url()).toBe(
+      '/find?q=&language_code=(%22es%22%20OR%20%22hi%22)');
+  });
+
+  it('should filter and search content by categories, language and text',
+    function() {
+      var calls = 0;
+      spyOn(searchService, 'getCurrentUrlQueryString').and.callFake(function() {
+        calls++;
+        if (calls === 1) {
```

Comment:
This is weird.... Can you try to do this in a cleaner fashion?

Reply:
I've tried to test this function but `getCurrentUrlQueryString` was always returning the same value. It must return different values to test the search-bar behavior so that's the way I've found to do it.

Reply:
Hmmm, I see. But it is still a bit confusing, can we make 2 separate mocks and use the one that is necessary in whichever scenario.

Reply:
I don't think it's going to work in this case. Using more than 1 mock in the same spec works well when there is an explicit action between the calls (for example, HTTP calls, `$rootScope` calls, and so on). In this case, both calls are called after building a query, which I'm mocking with promise, that doesn't call `then or` `catch`. So there is no explicit action to separate the two mocks.

Reply:
I have a suggestion. Instead of mocking `getCurrentUrlQueryString` can we have 2 mock outputs for `getSearchUrlQueryString` based on the arguments passed. Maybe we can do something like this: https://stackoverflow.com/a/16208124?

Reply:
`getCurrentUrlQueryString` doesn't take any arguments in the component, I think it's not possible to use this approach.

Reply:
Hi @marianazangrossi, I'm saying let's mock `getSearchUrlQueryString` (which is what `getCurrentUrlQueryString` calls), this method takes `_lastQuery`, `_lastSelectedCategories`, and `_lastSelectedLanguageCodes` as arguments.

Reply:
@kevintab95 I tried to mock `getSearchUrlQueryString` but it didn't work. I think it's a tricky situation because `_lastQuery`, `_lastSelectedCategories`, and `_lastSelectedLanguageCodes` are updated only after an HTTP get request, which doesn't happen between the first and second call of `getSearchUrlQueryString` (and `getCurrentUrlQueryString`). I don't see why `oldQuery` and `newQuery` are being used in this case, but I don't remove it because it's probably to handle a tricky situation that it's not clear in the code.

Reply:
> .. updated only after an HTTP get request, which doesn't happen between the first and second call of getSearchUrlQueryString (and getCurrentUrlQueryString) ..

Is it possible to execute the necessary actions in your test before the second call in order to update the parameter values? Sounds like making the GET call twice with the correct values should work.

Reply:
I'm not sure if it's going to work. When making the second HTTP call, both `getSearchUrlQueryString` will return the same value because they don't depend on any request. FYI, here's the original function:
```js
var updateSearchFieldsBasedOnUrlQuery = function() {
  var oldQueryString = SearchService.getCurrentUrlQueryString();

  // There is no HTTP call here to update the query string
  ctrl.selectionDetails.categories.selections = {};
  ctrl.selectionDetails.languageCodes.selections = {};

  ctrl.searchQuery =
    SearchService.updateSearchFieldsBasedOnUrlQuery(
      $window.location.search, ctrl.selectionDetails);

  updateSelectionDetails('categories');
  updateSelectionDetails('languageCodes');
  // There is no HTTP call here to update the query string

  var newQueryString = SearchService.getCurrentUrlQueryString();

  if (oldQueryString !== newQueryString) {
    onSearchQueryChangeExec();
  }
};
```

I don't understand the code itself (why is it getting the current URL query string if it is the same?) but I think it's to handle some specific tricky situation we're unaware.

Reply:
Oh I see.. Here's my understanding:

We watch the changes on `ctrl.searchQuery`, so when we change its value by assigning it `SearchService.updateSearchFieldsBasedOnUrlQuery(..)`, the watcher gets activated and executes the search. It calls `onSearchQueryChangeExec` which in turn calls `SearchService.executeSearchQuery()`.

https://github.com/oppia/oppia/blob/35573f3b92acc6628f14005d05728a5738690985/core/templates/pages/library-page/search-bar/search-bar.directive.ts#L243-L249

But I guess the question then becomes, why do `onSearchQueryChangeExec` again inside `updateSearchFieldsBasedOnUrlQuery`. I'm not sure.

Reply:
I have the same question as yours. For one moment I thought removing the lines from `updateSearchFieldsBasedOnUrlQuery` would work but I was afraid to break something in the application. Do you have a suggestion of who can I talk to about this problem?

Reply:
Looks like the logic in `updateSearchFieldsBasedOnUrlQuery` was introduced in https://github.com/oppia/oppia/pull/1686 but the watcher was added much later in https://github.com/1995YogeshSharma/oppia/commit/f639768bd9b1e21495d25589ab347525a356fa91#diff-4158b82cd999a3be1f0992fba9812da6. You can check with the authors to see if the redundant code in `updateSearchFieldsBasedOnUrlQuery` can be removed.

Reply:
Thanks for looking up the PRs. @kevinlee12 and @seanlip, can you guys please help me clarify this problem, since you're mentioned in the PRs? Do you have a suggestion for it?

Reply:
I think what happened is that the author of #3006 didn't notice that some code was redundant after adding the $watch. This is one of the problems with $watch, though -- it makes it hard to trace dependencies.

My suggestion would be to get rid of the $watch altogether, and in every action that changes the searchQuery variable, replace the assignment with the following (note: pseudocode)

```
if new thing !== current search query:
    set search query to new thing
    call onSearchQueryChangeExec()
```

Would that work?

Reply:
Thanks for the explanation @seanlip! After doing some research, I've done some changes in the code. I decided to keep `$watch` and remove the code from `updateSearchFieldsBasedOnUrlQuery` method because it didn't add any logic into the component. The `newQuery` and `oldQuery` would always be equal because there is no change happening between the two calls of `getCurrentUrlQueryString`. In the `$watch` method, when `ctrl.searchQuery` is changed, it will get fetch the data from backend again.

I tested the component's behavior and it seems it's working perfectly. However, do you guys know how to test it by steps to be sure it's 100% correct?

Reply:
But $watch is an antipattern, as I mentioned above. The refactor I suggested doesn't add logic to the component, it just moves the existing logic into a different place.

I'm not sure we can give you steps to test the whole component. Maybe you can list what you've done and we can tell you if we think there are omissions?

Reply:
I see... I've tried to remove `$watch` and make the changes you suggested. However, search-bar breaks without the $watch, it's not possible to search anymore, here's a gif:
![file3](https://user-images.githubusercontent.com/34922478/90822038-336a3d00-e30a-11ea-967c-95289d842aee.gif)

Reply:
I see... I removed the $watch method and I followed the suggestions you've made. However, now search-bar is breaking, I can search by text anymore. Here's a gif:
![file4](https://user-images.githubusercontent.com/34922478/90822459-e0dd5080-e30a-11ea-9a73-feb50006b9cd.gif)

Reply:
Ah -- you need to add a listener in the ng-model too in the directive, I think. Try something like ng-change, see if that works? https://docs.angularjs.org/api/ng/directive/ngChange

Reply:
Thanks a lot! I think it worked!